compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGenOptimizationReportEmitter.java

```diff
@@ -0,0 +1,172 @@
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.TARGET_DIR_NAME;
+
+/**
+ * Emits time durations and optimized node details for codegen optimization.
+ *
+ * @since 2201.10.0
+ */
+public class CodeGenOptimizationReportEmitter {
+
+    private static final Map<PackageID, Long> birOptimizationDurations = new LinkedHashMap<>();
+    private static long nativeOptimizationDuration = 0;
+    private static final PrintStream out = System.out;
+    private static final String CODEGEN_OPTIMIZATION_REPORT = "codegen_optimization_report.json";
+
+    protected static void flipBirOptimizationTimer(PackageID pkgId) {
+        if (birOptimizationDurations.containsKey(pkgId)) {
+            long totalDuration = System.currentTimeMillis() - birOptimizationDurations.get(pkgId);
+            birOptimizationDurations.put(pkgId, totalDuration);
+        } else {
+            birOptimizationDurations.put(pkgId, System.currentTimeMillis());
+        }
+    }
+
+    protected static void flipNativeOptimizationTimer() {
+        if (nativeOptimizationDuration == 0) {
+            nativeOptimizationDuration = System.currentTimeMillis();
+        } else {
+            nativeOptimizationDuration = System.currentTimeMillis() - nativeOptimizationDuration;
+        }
+    }
+
+    protected static void emitBirOptimizationDuration() {
+        long totalDuration = birOptimizationDurations.values().stream().mapToLong(Long::longValue).sum();
+        out.printf("Duration for unused BIR node analysis : %dms%n", totalDuration);
+        birOptimizationDurations.forEach((key, value) -> out.printf("    %s : %dms%n", key, value));
+        out.println();
+        // TODO make this class a singleton
+        birOptimizationDurations.clear();
+    }
+
+    protected static void emitNativeOptimizationDuration() {
+        out.println(
+                "Duration for Bytecode Optimization (analysis + deletion) : " + nativeOptimizationDuration + "ms");
+        nativeOptimizationDuration = 0;
+    }
+
+    protected static void emitOptimizedExecutableSize(Path executableFilePath) {
+        try {
+            float optimizedJarSize = Files.size(executableFilePath) / (1024f * 1024f);
+            System.out.printf("Optimized file size : %f MB%n", optimizedJarSize);
+        } catch (IOException e) {
+            throw new ProjectException("Failed to emit optimized executable size ", e);
+        }
+    }
+
+    protected static void emitCodegenOptimizationReport(
+            Map<PackageID, UsedBIRNodeAnalyzer.InvocationData> invocationDataMap, Path projectDirectoryPath,
+            ProjectKind projectKind) {
+        Path reportParentDirectoryPath = projectDirectoryPath.resolve(TARGET_DIR_NAME).toAbsolutePath().normalize();
```

Comment:
Users can provide a custom path for the target directory. We can use `Project.targetDir()` which handles this and returns the correct target path.

Reply:
We resorted to this way because for single file projects this seems to give a generated temporary directory,

![image](https://github.com/ballerina-platform/ballerina-lang/assets/105668612/efb46652-19ba-4a3d-94b8-fbd81d3be992)

But if we use `project().sourceRoot.toAbsolutePath()`, it gives the correct location of the project,

![image](https://github.com/ballerina-platform/ballerina-lang/assets/105668612/d301f334-7754-4d29-8c53-e4a0bfbc7e54)

Is there a way to get the correct target path for single file projects using `Project.targetDir()`?

Reply:
After a discussion with @gayaldassanayake, we decided to use `Project.targetDir()` for both normal and single file projects. After this change, if users need to generate optimization reports for single file projects, they will have to provide a custom target directory. I will update here after refactoring.

Reply:
After a discussion with @azinneera, we decided to manually resolve the optimization report path only for single-file projects. For build projects `Project.targetDir()` will be used as the default optimization report path. Fix can be found in c2811f5f6a44844c14f04bba54927f8bdfbd5838.